crates/hir_ty/src/infer/expr.rs

```diff
@@ -738,7 +738,8 @@ impl<'a> InferenceContext<'a> {
                 let len = match array {
                     Array::ElementList(items) => {
                         for expr in items.iter() {
-                            let cur_elem_ty = self.infer_expr_inner(*expr, expected);
+                            let cur_elem_ty = self
+                                .infer_expr_inner(*expr, &Expectation::has_type(elem_ty.clone()));
```

Comment:
I think we should only use the original expectation, not the modified `elem_ty`

Reply:
Do you mean like this?

```diff
diff --git i/crates/hir_ty/src/infer/expr.rs w/crates/hir_ty/src/infer/expr.rs
index 71a1088b4..fea3264ec 100644
--- i/crates/hir_ty/src/infer/expr.rs
+++ w/crates/hir_ty/src/infer/expr.rs
@@ -734,12 +734,11 @@ impl<'a> InferenceContext<'a> {
                         Some(TyKind::Array(st, _) | TyKind::Slice(st)) => st.clone(),
                         _ => self.table.new_type_var(),
                     };
-
+                let expected = Expectation::has_type(elem_ty.clone());
                 let len = match array {
                     Array::ElementList(items) => {
                         for expr in items.iter() {
-                            let cur_elem_ty = self
-                                .infer_expr_inner(*expr, &Expectation::has_type(elem_ty.clone()));
+                            let cur_elem_ty = self.infer_expr_inner(*expr, &expected);
                             elem_ty = self.coerce_merge_branch(Some(*expr), &elem_ty, &cur_elem_ty);
                         }
                         Some(items.len() as u64)
```

Reply:
exactly